### PR TITLE
Add TLS support for trunk serve

### PR DESCRIPTION
**Pull request by lpotthast**
Cherry-picked the change from @lukaselmer. Gonna verify it with a project I am working on. 
Closes https://github.com/ctron/trunk/issues/4

----

```
Add TLS support for trunk serve

Use clippy, improve error handling

Improve tls example

Rename TLS options, improve path resolution

Move Rustls config parsing into rt.rs

# Conflicts:
#	CHANGELOG.md
#	Cargo.lock
#	Cargo.toml
#	Trunk.toml
#	src/config/models.rs
#	src/config/rt.rs
#	src/serve.rs
```

### CODE REVIEW

**Comment by lpotthast** [comment]
@ctron Works good on my end!

File: examples/yew-tls/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Trunk | Yew | ybc
```
Comment (ctron) [inline examples/yew-tls/README.md:1]: I think the example is more about TLS?

File: examples/yew-tls/self_signed_certs/cert.pem [hunk 2]
```diff
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
```
Comment (ctron) [inline examples/yew-tls/self_signed_certs/cert.pem:1]: The certificate seems to have expired 2022. I think for a certificate like this, having some expiry date in 2050 would work best :)

File: src/serve.rs [hunk 3]
```diff
@@ -126,10 +133,26 @@ impl ServeSystem {
         ));
         let router = router(state, cfg.clone())?;
         let addr = (cfg.address, cfg.port).into();
-        let server = Server::bind(&addr)
-            .serve(router.into_make_service())
-            .with_graceful_shutdown(shutdown_fut);
 
+        let mut http_server: Option<_> = None;
```
Comment (ctron) [inline src/serve.rs:137]: I am not a fan of this duplication … and I don't fully understand what's going on with the shutdown. I will try to see if we can improve this. Never worked with axum before, so I need to better understand this.

Comment (ctron) [reply src/serve.rs:137]: Just to be clear, I wouldn't consider this blocker. Just room for improvement.

**Review by ctron — CHANGES_REQUESTED** [review]
Thanks for the PR. I think it's good in general, some small things with the example which I would prefer to see fixed.

And the duplication issue (axum server start), but I think we'll figure this out.

File: src/serve.rs (hunk 3 above)
Comment (ctron) [reply src/serve.rs]: Ok, I pushed some changes, consolidating the changes/differences into some function. Seems like the types are a bit complex to get them into a box or something.

I will go ahead and try this locally (non-TLS), maybe you can do the same (with TLS). It that works fine for both of us, then I think this part of resolved.

**Comment by ctron** [comment]
Lints should also be fixed now.

**Comment by lpotthast** [comment]
Aah, one thing is annoying. When performing a `build`, the `Trunk.toml` config is parsed and validated the same way as when running the `serve` command. In the nature of this change though: Users will probably generate their own certificates, with each user using its own certificate authority. In my project, the certs dir only contains a .gitkeep. Certs would not get commited. When running `build` inside a pipeline, no certs would be present then and the build aborts because the files referenced inside `Trunk.toml` could not be found... Quick pipeline-based fix would be
```
sed -i '/tls_cert_path/d' "./Trunk.toml"
sed -i '/tls_key_path/d' "./Trunk.toml"
```
but that should never be necessary. I think Trunk should actually ignore the tls configuration when not performing a serve. But that would require some conditional configuration evaluation, which is not currently there as it seems..

**Comment by ctron** [comment]
But the certificates are only present in the `serve` configuration, does that really effect any other command?

**Comment by ctron** [comment]
Ok, just tried it out. It's really causing an error.

**Comment by ctron** [comment]
I think the issue comes from the fact that it tries to canonicalize the paths during reading the configuration. I am not sure that's a necessary step though. Just using the path as is should work fine.

**Comment by ctron** [comment]
I noticed that the paths get actually canonicalized twice. First when parsing the configuration (in any case), second when starting the server (serve only). TBH, I am not convinced it is necessary in either case. But it gives a better error message when files are missing.

However, I think it's enough to do this when starting the `serve` command only. I made that change and it looks good to me. Maybe you can double check.

**Review by ctron — APPROVED** [review]
I think I fixed the remaining issues. Once the CI passes, I am going to merge it.

**Pull request merged**